Extension/src/LanguageServer/Providers/callHierarchyProvider.ts

```diff
@@ -88,33 +88,52 @@ export enum CallHierarchyRequestStatus {
     new RequestType<CallHierarchyParams, CallHierarchyItemResult, void>('cpptools/prepareCallHierarchy');
 
+const CallHierarchyCallsToRequest: RequestType<CallHierarchyParams, CallHierarchyCallsItemResult, void> =
+    new RequestType<CallHierarchyParams, CallHierarchyCallsItemResult, void>('cpptools/callHierarchyCallsTo');
+
 const CallHierarchyCallsFromRequest: RequestType<CallHierarchyParams, CallHierarchyCallsItemResult, void> =
     new RequestType<CallHierarchyParams, CallHierarchyCallsItemResult, void>('cpptools/callHierarchyCallsFrom');
 
 export class CallHierarchyProvider implements vscode.CallHierarchyProvider {
     // Indicates whether a request is from an entry root node (e.g. top function in the call tree).
     private isEntryRootNodeTelemetry: boolean = false;
     private client: DefaultClient;
+
     constructor(client: DefaultClient) {
         this.client = client;
     }
 
     public async prepareCallHierarchy(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken):
         Promise<vscode.CallHierarchyItem | undefined> {
+        await this.client.requestWhenReady(() => processDelayedDidOpen(document));
+        workspaceReferences.cancelCurrentReferenceRequest(CancellationSender.NewRequest);
         workspaceReferences.clearViews();
+
         const range: vscode.Range | undefined = document.getWordRangeAtPosition(position);
         if (range === undefined) {
             return undefined;
         }
 
-        await this.client.requestWhenReady(() => processDelayedDidOpen(document));
+        // Listen to a cancellation for this request. When this request is cancelled,
+        // use a local cancellation source to explicitly cancel a token.
+        const cancelSource: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+        const cancellationTokenListener: vscode.Disposable = token.onCancellationRequested(() => {
+            cancelSource.cancel();
+        });
+        const requestCanceledListener: vscode.Disposable = workspaceReferences.onCancellationRequested(sender => {
+            cancelSource.cancel();
+        });
 
         const params: CallHierarchyParams = {
             textDocument: { uri: document.uri.toString() },
             position: Position.create(position.line, position.character)
         };
-        const response: CallHierarchyItemResult = await this.client.languageClient.sendRequest(CallHierarchyItemRequest, params, token);
-        if (token.isCancellationRequested || response.succeeded === undefined) {
+        const response: CallHierarchyItemResult = await this.client.languageClient.sendRequest(CallHierarchyItemRequest, params, cancelSource.token);
+
+        cancellationTokenListener.dispose();
+        requestCanceledListener.dispose();
+
+        if (cancelSource.token.isCancellationRequested || response.succeeded === undefined) {
             throw new vscode.CancellationError();

```

Comment:
This throw causes VS Code to show 
![image](https://github.com/microsoft/vscode-cpptools/assets/19859882/9042dfa2-0fa9-4ced-871b-90afc2b1c5e2)
so I think it should be changed to return undefined, although a bug should be logged for them to fix that -- I haven't located the spot in their code which is triggering the error message.

Reply:
Fixed.